File: src/commands/audio-transcribe.ts
import { Credentials } from '../credentials';
import { RefreshDeps } from '../refresh';
import { getAccessToken } from '../auth';
import {
  AudioInputDeps,
  resolveAudioInput,
  writeApiErrorAndExit,
} from '../audio-io';
import { ENDPOINTS } from '../endpoints';
import { ENTRYPOINT_HEADER } from '../entrypoint';

export interface AudioTranscribeOptions {
  model?: string;
  language?: string;
  prompt?: string;
  responseFormat?: string;
  temperature?: string;
  filename?: string;
  contentType?: string;
}

export interface AudioTranscribeDeps extends AudioInputDeps {
  credentials?: Credentials;
  refreshDeps?: RefreshDeps;
  fetch?: typeof globalThis.fetch;
  now?: () => Date;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
}

export async function audioTranscribeCommand(
  fileArg: string | undefined,
  options: AudioTranscribeOptions,
  deps?: AudioTranscribeDeps,
): Promise<void> {
  const fetchFn = deps?.fetch ?? globalThis.fetch;
  const stdout = deps?.stdout ?? process.stdout;
  const stderr = deps?.stderr ?? process.stderr;

  const { token, routerUrl } = await getAccessToken(deps);

  const { data, filename, contentType } = await resolveAudioInput(
    fileArg,
    options,
    deps ?? {},
  );

  const form = new FormData();
  form.append('file', new Blob([data], { type: contentType }), filename);
  if (options.model) form.append('model', options.model);
  if (options.language) form.append('language', options.language);
  if (options.prompt) form.append('prompt', options.prompt);
  if (options.responseFormat) form.append('response_format', options.responseFormat);
  if (options.temperature) form.append('temperature', options.temperature);

  const response = await fetchFn(`${routerUrl}${ENDPOINTS.audioTranscriptions}`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}`, ...ENTRYPOINT_HEADER },
    body: form,
  });

  if (!response.ok) await writeApiErrorAndExit(response as any, stderr);

  const body = await response.text();
  stdout.write(body);
  if (!body.endsWith('\n')) stdout.write('\n');
}

File: src/commands/audio_transcribe.rs
use std::io::Write;

use crate::audio_io::{resolve_audio_input, write_api_error_and_exit};
use crate::auth::get_access_token;
use crate::credentials::Credentials;
use crate::endpoints;
use crate::error::Result;
use crate::net::{Client, FilePart};
use crate::origin::origin_header;

#[derive(Default)]
pub struct AudioTranscribeOptions {
    pub model: Option<String>,
    pub language: Option<String>,
    pub prompt: Option<String>,
    pub response_format: Option<String>,
    pub temperature: Option<String>,
    pub filename: Option<String>,
    pub content_type: Option<String>,
}

pub fn audio_transcribe_command(
    client: &Client,
    creds: &Credentials,
    file_arg: Option<&str>,
    options: AudioTranscribeOptions,
) -> Result<()> {
    let session = get_access_token(client, creds);

    let input = resolve_audio_input(
        file_arg,
        options.filename.as_deref(),
        options.content_type.as_deref(),
    )?;

    let mut fields: Vec<(&str, &str)> = Vec::new();
    if let Some(v) = &options.model {
        fields.push(("model", v));
    }
    if let Some(v) = &options.language {
        fields.push(("language", v));
    }
    if let Some(v) = &options.prompt {
        fields.push(("prompt", v));
    }
    if let Some(v) = &options.response_format {
        fields.push(("response_format", v));
    }
    if let Some(v) = &options.temperature {
        fields.push(("temperature", v));
    }

    let bearer = format!("Bearer {}", session.token);
    let resp = client.post_multipart(
        &format!("{}{}", session.router_url, endpoints::AUDIO_TRANSCRIPTIONS),
        &[("Authorization", &bearer), origin_header()],
        FilePart {
            field: "file".to_string(),
            filename: input.filename,
            content_type: input.content_type,
            data: input.data,
        },
        &fields,
    )?;

    if !resp.ok() {
        write_api_error_and_exit(&resp);
    }

    let body = resp.text();
    let mut out = std::io::stdout();
    out.write_all(body.as_bytes())?;
    if !body.ends_with('\n') {
        out.write_all(b"\n")?;
    }
    Ok(())
}
